### ISSUE TITLE
Confusing reborrow hints when calling immutable method from mutable method

### ISSUE THREAD

**Issue by al2me6** [rust-lang/rust-analyzer#13131]
**rust-analyzer version**: 0.4.1183-standalone (https://github.com/rust-lang/rust-analyzer/commit/6bea872edd9523a06213270f68725c9fe33f3919 2022-08-26)

**rustc version**: rustc 1.65.0-nightly (878aef79d 2022-08-20)

**relevant settings**: `"rust-analyzer.inlayHints.reborrowHints.enable": "always"`

For the snippet 
```rust
struct Foo;

impl Foo {
    fn immutable(&self) -> usize {
        0
    }

    fn mutable(&mut self) {
        let x = self.immutable();
    }
}
```

r-a gives the hint

```rust
let x = &*self.immutable();
//      ^^
```

which doesn't compile.

When `self` is manually parenthesized, r-a does show the correct hint:

```rust
let x = (&*self).immutable();
//       ^^
```

**Comment by flodiebold** [comment]
The hints aren't necessarily syntactically correct anyway, and I don't think we want to add parentheses to the hints, so I wouldn't consider this a bug.

**Comment by Veykril** [comment]
Tbf, I think I forgot about these rendering auto-derefs. They were mostly meant for visualizing mutable reborrows in the first place (that was the main magic I was thinking of), so they aren't really polished. Imo it would be great if we could visalize all the compiler inserted type adjustments (with proper syntax) as that makes for a great learning tool in my eyes. So in my opinion for those, we should actually insert parentheses where needed.

> (If we could find some generally nicer way to communicate this information, that would of course be nice. That's why these hints aren't enabled by default.)

We should probably improve the docs for the config at least.

These would nevertheless stay disabled by default since they are pretty noisy.

**Comment by WaffleLapkin** [comment]
I believe this has been fixed by #13816.

Current output (note that the inlay hint contains parens):

![Image](https://github.com/user-attachments/assets/125fd5bc-e190-4d53-9ed6-71406e8e084c)
![Image](https://github.com/user-attachments/assets/1d423390-14d1-446c-be62-1d0f27832d4b)
